stablehlo/reference/Ops.cpp

```diff
@@ -171,6 +171,27 @@ SmallVector<InterpreterValue> eval(
       auto inputs = scope.findTokens(afterAllOp.getInputs());
       auto result = evalAfterAllOp(inputs, afterAllOp->getContext());
       scope.add(afterAllOp.getResult(), result);
+    } else if (auto allReduceOp = dyn_cast<AllReduceOp>(op)) {
+      auto operand = scope.findTensor(allReduceOp.getOperand());
+
+      auto replicaGroupsAttr = allReduceOp.getReplicaGroups();
+      auto replicaGroupsShape = replicaGroupsAttr.getShapedType().getShape();
+      SmallVector<SmallVector<uint32_t>> replicaGroups(
+          replicaGroupsAttr.getNumElements() / replicaGroupsShape[1]);
+      for (auto &group : replicaGroups) {
+        auto it = replicaGroupsAttr.getValues<int64_t>().begin();
+        for (auto i = 0; i < replicaGroupsShape[1]; ++i) group.push_back(*it++);
+      }
```

Comment:
We'll need to change this code because AllReduceOp allows replica groups of different sizes. Since DenseElementsAttr doesn't support ragged tensors, we're currently using a convention where elements with value -1 represents raggedness (see http://google3/third_party/tensorflow/compiler/xla/translate/mhlo_to_hlo/attribute_exporter.cc;l=102-107;rcl=554330020).